gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java

```diff
@@ -147,33 +181,49 @@ protected void createMetrics() {
 
   @Override
   protected void processMessage(DecodeableKafkaRecord<byte[],byte[]> message) {
-    try {
-      GobblinTrackingEvent gobblinTrackingEvent = deserializeEvent(message);
+    GobblinTrackingEvent gobblinTrackingEvent = deserializeEvent(message);
 
-      if (gobblinTrackingEvent == null) {
-        return;
-      }
+    if (gobblinTrackingEvent == null) {
+      return;
+    }
 
-      if (IssueEventBuilder.isIssueEvent(gobblinTrackingEvent)) {
-        try (Timer.Context context = getMetricContext().timer(PROCESS_JOB_ISSUE).time()) {
-          jobIssueEventHandler.processEvent(gobblinTrackingEvent);
-        }
+    if (IssueEventBuilder.isIssueEvent(gobblinTrackingEvent)) {
+      try (Timer.Context context = getMetricContext().timer(PROCESS_JOB_ISSUE).time()) {
+        jobIssueEventHandler.processEvent(gobblinTrackingEvent);
       }
+    }
 
-      if (gobblinTrackingEvent.getName().equals(JobEvent.WORK_UNITS_CREATED)) {
-        emitWorkUnitCountMetric(gobblinTrackingEvent);
-        return;
-      }
+    if (gobblinTrackingEvent.getName().equals(JobEvent.WORK_UNITS_CREATED)) {
+      emitWorkUnitCountMetric(gobblinTrackingEvent);
+      return;
+    }
 
-      org.apache.gobblin.configuration.State jobStatus = parseJobStatus(gobblinTrackingEvent);
-      if (jobStatus != null) {
-        try(Timer.Context context = getMetricContext().timer(GET_AND_SET_JOB_STATUS).time()) {
-          addJobStatusToStateStore(jobStatus, this.stateStore);
+    try {
+      persistJobStatusRetryer.call(() -> {
```

Comment:
I think we should retry all previous statements in this method as well, like deserializing an event(that can fail if kafka schema registry is broken), or saving issues, which can cause trouble if Mysql is down.

Reply:
I looked at that, but grew concerned we might retry the wrong things.  for instance, if a corrupted message somehow lacked the frame-aligning magic byte:
```
  public Schema readSchemaVersioningInformation(DataInputStream inputStream)
      throws IOException {
    if (inputStream.readByte() != KafkaAvroSchemaRegistry.MAGIC_BYTE) {
      throw new IOException("MAGIC_BYTE not found in Avro message.");
    }
```
we'd retry, although that would never change.  of the two solutions:
  a. to throw a more specific exception for that condition, so retry would accept the attempt and stop
  b. localize retrying (separately) around that particular invocation of the schema registry
... I vote for the latter.

of the other two sub-tasks:
  1. `JobIssueEventHandler.processEvent` doesn't unwind errors to the caller, so no indication of failure within.  while there could be correlation between (suppressed) failure there and later failure with the state store, I wondered whether that wouldn't be too presumptuous... and I was unclear whether `processEvent` is idempotent.
  2. work unit metrics recording only seems to throw on caller error (IllegalArg) and given potentially time-series metrics, I again wasn't clear on idempotence.

Reply:
Even if it's kafka deserialization problem, I think it could be beneficial to spam this error in logs repeatedly, so it's easier to find in logs(we just need to look at the recent errors). Also, those implementations can change over time. 

It kind of feels that if any of the kafka processing thread dies, the service will be in the state that it should never be in, and it will be silent about it. Maybe those retries should happen at even higher level, the entry point of the thread.

Reply:
AFA kafka deserialization erorrs: at present, such bad messages are skipped, not retried, which is what I believe is best.  for the specific case of failure to contact the schema registry, we could retry there too... although I thought a separate effort is underway to draw from a previously cached value of schema IDs...

overall, we may differentiate between retrying and silent failures.  let's restrict the former only to operations likely to succeed if retried.  when the thread dies (for whatever reason), we probably want to raise an externally monitorable condition, to alarm on.  that way it's not "silent".  if we don't already have that, we probably should create a separate ticket to track that monitoring enhancement.